src/net.cpp

```diff
@@ -894,40 +902,58 @@ static void EraseLastKElements(
-    // longest uptime overall. This helps protect tor peers, which tend to be otherwise
-    // disadvantaged under our eviction criteria.
-    const size_t initial_size = vEvictionCandidates.size();
-    size_t total_protect_size = initial_size / 2;
-    const size_t onion_protect_size = total_protect_size / 2;
-
-    if (onion_protect_size) {
-        // Pick out up to 1/4 peers connected via our onion service, sorted by longest uptime.
-        EraseLastKElements(vEvictionCandidates, CompareOnionTimeConnected, onion_protect_size,
-                           [](const NodeEvictionCandidate& n) { return n.m_is_onion; });
-    }
-
-    const size_t localhost_min_protect_size{2};
-    if (onion_protect_size >= localhost_min_protect_size) {
-        // Allocate any remaining slots of the 1/4, or minimum 2 additional slots,
-        // to localhost peers, sorted by longest uptime, as manually configured
-        // hidden services not using `-bind=addr[:port]=onion` will not be detected
-        // as inbound onion connections.
-        const size_t remaining_tor_slots{onion_protect_size - (initial_size - vEvictionCandidates.size())};
-        const size_t localhost_protect_size{std::max(remaining_tor_slots, localhost_min_protect_size)};
-        EraseLastKElements(vEvictionCandidates, CompareLocalHostTimeConnected, localhost_protect_size,
-                           [](const NodeEvictionCandidate& n) { return n.m_is_local; });
+    // To favorise the diversity of our peer connections, reserve up to half of
+    // these protected spots for onion, localhost and I2P peers, even if they're
+    // not longest uptime overall. This helps protect these higher-latency peers
+    // that tend to be otherwise disadvantaged under our eviction criteria.
+    const size_t initial_size = candidates.size();
+    const size_t total_protect_size{initial_size / 2};
+
+    // Protect up to 25% of the candidates by disadvantaged network.
+    const size_t max_protect_by_network{total_protect_size / 2};
+    size_t num_protected{0};
+
+    while (num_protected < max_protect_by_network) {
+        const size_t remaining_to_protect{max_protect_by_network - num_protected};
+        const size_t protect_per_network{std::max(remaining_to_protect / networks.size(), static_cast<size_t>(1))};
+        bool protected_at_least_one{false};
+        for (const Net& n : networks) {
+            if (n.count == 0) continue;
+            const size_t before = candidates.size();
+            EraseLastKElements(candidates, CompareNodeNetworkTime(n.is_local, n.id), protect_per_network,
+                               [&n](const NodeEvictionCandidate& c) { return n.is_local ? c.m_is_local : c.m_network == n.id; });
+            const size_t after = candidates.size();
+            if (before > after) {
+                protected_at_least_one = true;
+                num_protected += before - after;
+                if (num_protected >= max_protect_by_network) break;
```

Comment:
Following diff doesn't break `net_peer_eviction_tests`, is this an expected behavior ?

```
@@ -943,7 +943,7 @@ void ProtectEvictionCandidatesByRatio(std::vector<NodeEvictionCandidate>& candid
             if (before > after) {
                 protected_at_least_one = true;
                 num_protected += before - after;
-                if (num_protected >= max_protect_by_network) break;
+                if (num_protected == max_protect_by_network) break;
             }
         }
         if (!protected_at_least_one) break;
 ```

Reply:
Agree, the tests pass with this change.
```diff
-                if (num_protected >= max_protect_by_network) {
+                assert(num_protected <= max_protect_by_network);
+                if (num_protected == max_protect_by_network) {
```
I think it's expected behavior based on these lines:
```cpp
    while (num_protected < max_protect_by_network) {
        const size_t disadvantaged_to_protect{max_protect_by_network - num_protected};
        const size_t protect_per_network{std::max(disadvantaged_to_protect / networks.size(), static_cast<size_t>(1))};
```

Reply:
Given that `num_protected` is incremented with a non-constant number, that could be larger than `1`, I assume it could become larger than `max_protect_by_network` at some point. Thus `>=` for comparison instead of `==`.

How could replacing `>=` with `==` and an assert change anything? I think if it changes anything then it must be a trigger of the assert - if it was equal before, then behavior is unchanged (`>=` is the same as `==`) and if it was strictly larger, before we would have executed `break;` and after the change - triggered the assert.

To me that assert looks scary, even if some (non-obvious) arithmetic a few lines above mandates it will never be triggered. The code may be altered in the future which may change this. Also, even if it happens that `num_protected > max_protect_by_network`, then nothing really bad has happened - we protected some peers (more than we intended), the program can continue executing just fine.

_(the code at the time I write this does not contain an assert (a457f34e5039b75ee015b273028c3ee153656d5c))_

Reply:
Right, the assert was just for sanity testing (I wrote more tests but didn't find a way to hit the assert), no plan to add it.